bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+      }
+
+    /* connect ports */
+    if (error == 0) // we may want to make auto connect configurable (so it can be turned off)
+      {
+        std::map<std::string, DeviceDetails> devices = query_jack_devices (jack_client_);
+        std::map<std::string, DeviceDetails>::const_iterator di;
+
+        di = devices.find (devid_);
+        if (di != devices.end())
+          {
+            const DeviceDetails &details = di->second;
+
+            for (uint ch = 0; ch < n_channels_; ch++)
+              {
+                if (details.output_ports > ch)
+                  jack_connect (jack_client_, details.output_port_names[ch].c_str(),
+                                              jack_port_name (input_ports_[ch]));
+                if (details.input_ports > ch)
+                  jack_connect (jack_client_, jack_port_name (output_ports_[ch]),
+                                              details.input_port_names[ch].c_str());
+              }
+          }
+      }
+
+    /* setup PCM handle or shutdown */
+    if (error == 0)
+      {
+        flags_ |= Flags::OPENED;
+
+        uint dummy;
+        pcm_latency (&dummy, &dummy);   // debugging only: print latency values
+      }
+    else
+      {
+        disconnect_jack (jack_client_);
+        jack_client_ = nullptr;
+      }
+    JDEBUG ("JACK: opening PCM \"%s\" readupble=%d writable=%d: %s", devid_.c_str(), readable(), writable(), bse_error_blurb (error));
+    return error;
+  }
+  virtual bool
+  pcm_check_io (long *timeoutp) override
+  {
+    assert_return (jack_client_ != nullptr, false);
+
+    if (0)
+      {
+        /* One of the things that we want to test is whether the JACK driver would
+         * recover properly if a dropout should occur.
+         *
+         * Since dropouts occur rarely, we can use the TEST_DROPOUT macro. This
+         * will check if /tmp/drop exists, and if so, sleep for some time to
+         * ensure BSE can't write to/read from the ring buffer in time. Such an
+         * artificial dropout can be created using
+         *
+         * $ touch /tmp/drop
+         *
+         * The file will be removed and a sleep will happen.
+         *
+         *  - for production builds, the macro should never be used
+         *  - the macro invocation can be moved to other places in the source code
+         *    to introduce a dropout there
+         */
+        TEST_DROPOUT();
+      }
+
+    /* enable processing in callback (if not already active) */
+    atomic_active_ = 1;
+
+    /* report jack driver xruns */
+    if (atomic_xruns_ != printed_xruns_)
+      {
+        printed_xruns_ = atomic_xruns_;
+        Bse::printerr ("JACK: %d beast driver xruns\n", printed_xruns_);
+      }
+    /* report jack shutdown */
+    if (is_down_ && !printed_is_down_)
+      {
+        printed_is_down_ = true;
+        Bse::printerr ("JACK: connection to jack server lost\n");
+        Bse::printerr ("JACK:  -> to continue, manually stop playback and restart\n");
```

Comment:
Please add the devid_ to all printerr calls, I've also fixed this in the ALSA driver.

Reply:
Fixed.